include/xtensor/xreducer.hpp

```diff
@@ -37,20 +37,231 @@ namespace xt
+
+        xt::xarray<result_type> result;
+
+        if (!std::is_sorted(axes.cbegin(), axes.cend()))
+        {
+            throw std::runtime_error("Reducing axes should be sorted");
+        }
+
+        if (e.dimension() == axes.size())
+        {
+            auto begin = e.data().begin();
+            result_type tmp = init_fct(*begin);
+            ++begin;
+            result(0) = std::accumulate(begin, e.data().end(), tmp, acc_fct);
+            return result;
+        }
+
+        std::size_t idx = 0;
+        for (std::size_t i = 0; i < e.dimension(); ++i)
+        {
+            if (std::find(axes.begin(), axes.end(), i) == axes.end())
+            {
+                // i not in axes!
+                result_shape[idx] = e.shape()[i];
+                ++idx;
+            }
+        }
+
+        result.reshape(result_shape);
+
+        std::size_t inner_loop_size = e.strides()[axes[axes.size() - 1]];
+        std::size_t inner_stride    = e.strides()[axes[axes.size() - 1]];
+        std::size_t outer_loop_size = e.shape()[axes[axes.size() - 1]];
+
+        auto it = axes.rbegin();
+        auto last_ax = *it;
+        ++it;
+        for (; it != axes.rend(); ++it)
+        {
+            if (*it == last_ax - 1)
+            {
+                last_ax = *it;
+                outer_loop_size *= e.shape()[last_ax];
+            }
+        }
+
+        idx = 0;
+        for (std::size_t i = 0; i < e.dimension(); ++i)
+        {
+            if (std::find(axes.begin(), axes.end(), i) == axes.end())
+            {
+                // i not in axes!
+                iter_strides[i] = result.strides()[idx];
+                ++idx;
+            }
+        }
+
+        // TODO remove all entries where iter_shape == 0
+        iter_shape.erase(iter_shape.begin() + last_ax, iter_shape.end());
+        iter_strides.erase(iter_strides.begin() + last_ax, iter_strides.end());
+
+        for (std::size_t i = 0; i < iter_shape.size(); ++i)
+        {
+            if (iter_shape[i] == 0)
+            {
+                iter_shape.erase(iter_shape.begin() + i);
+                iter_strides.erase(iter_strides.begin() + i);
+                ++i;
+            }
+        }
+
+        xindex temp_idx(iter_shape.size());
+        auto next_idx = [&iter_shape, &iter_strides, &temp_idx]()
+        {
+            int i = iter_shape.size() - 1;
+            for (; i >= 0; --i)
+            {
+                if (ptrdiff_t(temp_idx[(std::size_t) i]) >= ptrdiff_t(iter_shape[(std::size_t) i]) - 1)
+                {
+                    temp_idx[(std::size_t) i] = 0;
+                }
+                else
+                {
+                    temp_idx[(std::size_t) i]++;
+                    break;
+                }
+            }
+            return std::make_pair(i == -1,
+                                  std::inner_product(temp_idx.begin(), temp_idx.end(),
+                                                     iter_strides.begin(), ptrdiff_t(0)));
```

Comment:
Ok the vector of pair may be annoying here; you need an additional `xindex` before the lambda definition to store an intermediate result; since we removed a dynamic allocation when we merged `iter_shape` and `iter_strides`, this won't be a performance issue compared to the initial implementation. Let's call it `tmp_prod`, you you would have to replace `std::inner_product` with:
```cpp
std::transform(temp_idx.begin(), temp.idx.end(), iter_shape_strides.begin(), tmp_prod.begin(), auto[](auto&& t1, auto&& t2) { return t1 * t2.second; });
return std::make_pair(i == -1,
std::accumulate(tmp_prod.begin(), tmp_prod.end(), ptrdiff_t(0)));
```
This less expressive than `inner_product`, but it is inside a lambda whose name is expressive enough, and it simplifies the code before so I think it is worth doing it.

Reply:
I think there is a more straightforward way to initialize the iter_shape & iter_strides, without having to remove any elements by just filling in the right ones. But I kind of had to go back to the drawing board with this function and derive it logically before making such optimizations.

I'll try to add that later in this PR.